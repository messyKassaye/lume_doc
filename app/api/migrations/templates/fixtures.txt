import db from 'api/utils/testing_db';

export default {
  entities: [{ title: 'test_doc' }]
};
