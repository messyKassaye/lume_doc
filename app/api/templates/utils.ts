import uuid from 'node-uuid';
import { PropertySchema } from 'shared/types/commonTypes';
import settings from 'api/settings/settings';
import { ThesaurusValueSchema } from 'shared/types/thesaurusType';

const generateNewSafeName = (label: string) =>
  label
    .trim()
    .replace(/[#|\\|/|*|?|"|<|>|=|||\s|:|.]/gi, '_')
    .replace(/^[_|\-|+|$]/, '')
    .toLowerCase();

export const safeName = (label: string, newNameGeneration: boolean = false) => {
  if (newNameGeneration) {
    return generateNewSafeName(label);
  }
  return label
    .trim()
    .replace(/[^a-z0-9]/gi, '_')
    .toLowerCase();
};

const generateName = (property: PropertySchema, newNameGeneration: boolean) => {
  const name = property.label ? safeName(property.label, newNameGeneration) : property.name;
  return property.type === 'geolocation' || property.type === 'nested'
    ? `${name}_${property.type}`
    : name;
};

const generateNames = (properties: PropertySchema[], newNameGeneration: boolean) =>
  properties.map(property => ({
    ...property,
    name: generateName(property, newNameGeneration),
  }));

export function generateIds(properties: PropertySchema[] = []) {
  return properties.map(property => ({
    ...property,
    id: property.id || uuid.v4(),
  }));
}

export const generateNamesAndIds = async (properties: PropertySchema[] = []) => {
  const { newNameGeneration = false } = await settings.get();
  return generateIds(generateNames(properties, newNameGeneration));
};

export interface PropertyOrThesaurusSchema
  extends Partial<PropertySchema>,
    Partial<ThesaurusValueSchema> {}

const flattenProperties = (properties: PropertyOrThesaurusSchema[]) =>
  properties.reduce<PropertyOrThesaurusSchema[]>((flatProps, p) => {
    if (p.values) {
      return [...flatProps, ...p.values];
    }

    return [...flatProps, p];
  }, []);

export function getUpdatedNames(
  oldProperties: PropertyOrThesaurusSchema[] = [],
  newProperties: PropertyOrThesaurusSchema[],
  prop: 'name' | 'label' = 'name',
  outKey: 'name' | 'label' = prop
) {
  const propertiesWithNewName: { [k: string]: string | undefined } = {};

  flattenProperties(oldProperties).forEach(property => {
    const newProperty = flattenProperties(newProperties).find(p => p.id === property.id);
    if (newProperty && newProperty[prop] !== property[prop]) {
      const key = property[outKey];
      if (key) {
        propertiesWithNewName[key] = newProperty[prop];
      }
    }
  });

  return propertiesWithNewName;
}

const notIncludedIn = (propertyCollection: PropertyOrThesaurusSchema[]) => (
  property: PropertyOrThesaurusSchema
) => !propertyCollection.find(p => p.id === property.id);

export function getDeletedProperties(
  oldProperties: PropertyOrThesaurusSchema[] = [],
  newProperties: PropertyOrThesaurusSchema[],
  prop: 'name' | 'label' = 'name'
) {
  return flattenProperties(oldProperties)
    .filter(notIncludedIn(flattenProperties(newProperties)))
    .map(property => property[prop]);
}
