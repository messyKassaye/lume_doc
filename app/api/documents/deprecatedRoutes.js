import Joi from 'joi';

import { validation } from '../utils';
import documents from './documents';
import needsAuthorization from '../auth/authMiddleware';
import templates from '../templates';

export default app => {
  app.post('/api/documents', needsAuthorization(['admin', 'editor']), (req, res, next) =>
    documents
      .save(req.body, { user: req.user, language: req.language })
      .then(doc => res.json(doc))
      .catch(next)
  );

  app.get(
    '/api/documents/count_by_template',
    validation.validateRequest(
      Joi.object()
        .keys({
          templateId: Joi.objectId().required(),
        })
        .required(),
      'query'
    ),
    (req, res, next) =>
      templates
        .countByTemplate(req.query.templateId)
        .then(results => res.json(results))
        .catch(next)
  );

  app.get(
    '/api/documents',
    validation.validateRequest(
      Joi.object().keys({
        _id: Joi.string().required(),
      }),
      'query'
    ),
    (req, res, next) => {
      let id;

      if (req.query && req.query._id) {
        id = req.query._id;
      }

      documents
        .getById(id, req.language)
        .then(response => {
          if (!response) {
            res.json({}, 404);
            return;
          }
          res.json({ rows: [response] });
        })
        .catch(next);
    }
  );

  app.delete(
    '/api/documents',
    needsAuthorization(['admin', 'editor']),
    validation.validateRequest(
      Joi.object({
        sharedId: Joi.string().required(),
      }).required(),
      'query'
    ),

    (req, res, next) => {
      documents
        .delete(req.query.sharedId)
        .then(response => res.json(response))
        .catch(next);
    }
  );
};
