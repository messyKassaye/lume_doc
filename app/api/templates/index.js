import * as templateUtils from './utils';

export { default } from './templates';

export { templateUtils };
