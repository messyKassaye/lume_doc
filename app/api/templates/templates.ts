import entities from 'api/entities';
import translations from 'api/i18n/translations';
import createError from 'api/utils/Error';
import { TemplateSchema } from 'shared/types/templateType';
import { PropertySchema } from 'shared/types/commonTypes';
import { updateMapping } from 'api/search/entitiesIndex';
import { ensure } from 'shared/tsUtils';
import { ObjectID } from 'mongodb';

import { validateTemplate } from '../../shared/types/templateSchema';
import model from './templatesModel';
import { generateNamesAndIds, getDeletedProperties, getUpdatedNames } from './utils';

const removePropsWithNonexistentId = async (nonexistentId: string) => {
  const relatedTemplates = await model.get({ 'properties.content': nonexistentId });
  await Promise.all(
    relatedTemplates.map(async t =>
      model.save({
        ...t,
        properties: (t.properties || []).filter(prop => prop.content !== nonexistentId),
      })
    )
  );
};

const createTranslationContext = (template: TemplateSchema) => {
  const titleProperty = ensure<PropertySchema>(
    ensure<PropertySchema[]>(template.commonProperties).find(p => p.name === 'title')
  );

  const context = (template.properties || []).reduce<{ [k: string]: string }>((ctx, prop) => {
    ctx[prop.label] = prop.label;
    return ctx;
  }, {});

  context[template.name] = template.name;
  context[titleProperty.label] = titleProperty.label;
  return context;
};

const addTemplateTranslation = async (template: TemplateSchema) =>
  translations.addContext(
    template._id,
    template.name,
    createTranslationContext(template),
    'Entity'
  );

const updateTranslation = async (currentTemplate: TemplateSchema, template: TemplateSchema) => {
  const currentProperties = currentTemplate.properties;
  const newProperties = template.properties || [];
  const updatedLabels = getUpdatedNames(currentProperties, newProperties, 'label');
  if (currentTemplate.name !== template.name) {
    updatedLabels[currentTemplate.name] = template.name;
  }
  const deletedPropertiesByLabel = getDeletedProperties(currentProperties, newProperties, 'label');
  const context = createTranslationContext(template);

  return translations.updateContext(
    currentTemplate._id,
    template.name,
    updatedLabels,
    deletedPropertiesByLabel,
    context,
    'Entity'
  );
};

export default {
  async save(template: TemplateSchema, language: string, reindex = true) {
    /* eslint-disable no-param-reassign */
    template.properties = template.properties || [];
    template.properties = await generateNamesAndIds(template.properties);
    /* eslint-enable no-param-reassign */

    await validateTemplate(template);

    await this.swapNamesValidation(template);

    if (reindex) {
      await updateMapping([template]);
    }

    if (template._id) {
      return this._update(template, language, reindex);
    }

    return model
      .save(template)
      .then(async newTemplate => addTemplateTranslation(newTemplate).then(() => newTemplate));
  },

  async swapNamesValidation(template: TemplateSchema) {
    if (!template._id) {
      return;
    }
    const current = await this.getById(ensure(template._id));

    const currentTemplate = ensure<TemplateSchema>(current);
    currentTemplate.properties = currentTemplate.properties || [];
    currentTemplate.properties.forEach(prop => {
      const swapingNameWithExistingProperty = (template.properties || []).find(
        p => p.name === prop.name && p.id !== prop.id
      );
      if (swapingNameWithExistingProperty) {
        throw createError(`Properties can't swap names: ${prop.name}`, 400);
      }
    });
  },

  async _update(template: TemplateSchema, language: string, reindex = true) {
    let _currentTemplate: TemplateSchema;
    return this.getById(ensure(template._id))
      .then(async current => {
        const currentTemplate = ensure<TemplateSchema>(current);
        return Promise.all([currentTemplate, updateTranslation(currentTemplate, template)]);
      })
      .then(([current]) => {
        const currentTemplate = ensure<TemplateSchema>(current);
        _currentTemplate = currentTemplate;
        const currentTemplateContentProperties = (currentTemplate.properties || []).filter(
          p => p.content
        );
        const templateContentProperties = (template.properties || []).filter(p => p.content);

        const toRemoveValues = currentTemplateContentProperties
          .map(prop => {
            const sameProperty = templateContentProperties.find(p => p.id === prop.id);
            if (sameProperty && sameProperty.content !== prop.content) {
              return sameProperty.name;
            }
            return null;
          })
          .filter(v => v);

        if (toRemoveValues.length === 0) {
          return;
        }
        return entities.removeValuesFromEntities(toRemoveValues, currentTemplate._id); // eslint-disable-line consistent-return
      })
      .then(async () => model.save(template))
      .then(async savedTemplate =>
        entities
          .updateMetadataProperties(template, _currentTemplate, language, reindex)
          .then(() => savedTemplate)
      );
  },

  async canDeleteProperty(template: ObjectID, property: ObjectID | string | undefined) {
    const tmps = await model.get();
    return tmps.every(iteratedTemplate =>
      (iteratedTemplate.properties || []).every(
        iteratedProperty =>
          !iteratedProperty.content ||
          !iteratedProperty.inheritProperty ||
          !(
            iteratedProperty.content.toString() === template.toString() &&
            iteratedProperty.inheritProperty.toString() === (property || '').toString()
          )
      )
    );
  },

  _validateSwapPropertyNames(currentTemplate: TemplateSchema, template: TemplateSchema) {
    (currentTemplate.properties || []).forEach(prop => {
      const swapingNameWithExistingProperty = (template.properties || []).find(
        p => p.name === prop.name && p.id !== prop.id
      );
      if (swapingNameWithExistingProperty) {
        throw createError(`Properties can't swap names: ${prop.name}`, 400);
      }
    });
  },

  async get(query: any = {}) {
    return model.get(query);
  },

  async setAsDefault(_id: string) {
    const [templateToBeDefault] = await this.get({ _id });
    const [currentDefault] = await this.get({ _id: { $nin: [_id] }, default: true });

    if (templateToBeDefault) {
      let saveCurrentDefault = Promise.resolve({});
      if (currentDefault) {
        saveCurrentDefault = model.save({
          _id: currentDefault._id,
          default: false,
        });
      }
      return Promise.all([model.save({ _id, default: true }), saveCurrentDefault]);
    }

    throw createError('Invalid ID');
  },

  async getById(templateId: ObjectID | string) {
    return model.getById(templateId);
  },

  async delete(template: TemplateSchema) {
    const count = await this.countByTemplate(ensure(template._id));
    if (count > 0) {
      return Promise.reject({ key: 'documents_using_template', value: count }); // eslint-disable-line prefer-promise-reject-errors
    }
    const _id = ensure<string>(template._id);
    await translations.deleteContext(_id);
    await removePropsWithNonexistentId(_id);
    await model.delete(_id);

    return template;
  },

  async countByTemplate(template: string) {
    return entities.countByTemplate(template);
  },

  async countByThesauri(thesauriId: string) {
    return model.count({ 'properties.content': thesauriId });
  },
};
