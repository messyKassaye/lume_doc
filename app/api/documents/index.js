import documents from './documents';

export default documents;
