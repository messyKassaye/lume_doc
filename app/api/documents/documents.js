import { createError } from 'api/utils';
import { files } from 'api/files';
import entities from '../entities';

const documents = {
  save(doc, params) {
    delete doc.file;
    return entities.save(doc, params);
  },

  async page(_id, page) {
    const [document] = await files.get({ _id }, '+fullText');
    if (!document || !document.fullText) {
      throw createError('document does not exists', 404);
    }

    if (typeof document.fullText[page] === 'undefined') {
      throw createError('page does not exists', 404);
    }

    const pageNumberMatch = /\[\[(\d+)\]\]/g;
    return document.fullText[page].replace(pageNumberMatch, '');
  },

  async savePDFInfo(_id, pdfInfo) {
    const [doc] = await files.get({ _id }, '+pdfInfo');
    if (doc.pdfInfo) {
      return doc;
    }
    const result = await files.save({ _id, pdfInfo });
    result.pdfInfo = pdfInfo;
    return result;
  },

  get(query, select) {
    return entities.get(query, select);
  },

  getById(sharedId, language) {
    return entities.getById(sharedId, language);
  },

  countByTemplate(templateId) {
    return entities.countByTemplate(templateId);
  },

  delete(id) {
    return entities.delete(id);
  },
};

export default documents;
export { documents };
