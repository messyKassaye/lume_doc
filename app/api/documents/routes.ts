import { Application, Request, Response, NextFunction } from 'express';

import { validation } from 'api/utils';
import { documents } from './documents';

export const documentRoutes = (app: Application) => {
  app.get(
    '/api/documents/page',

    validation.validateRequest({
      properties: {
        query: {
          properties: {
            _id: { type: 'string' },
            page: { type: 'string' },
          },
        },
      },
    }),

    async (req: Request, res: Response, next: NextFunction) => {
      documents
        .page(req.query._id, req.query.page)
        .then((result: string) => {
          res.json({ data: result });
        })
        .catch(next);
    }
  );
};
