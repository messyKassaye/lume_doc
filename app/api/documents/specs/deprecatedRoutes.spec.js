import { catchErrors } from 'api/utils/jasmineHelpers';
import db from 'api/utils/testing_db';

import documentRoutes from '../deprecatedRoutes.js';
import documents from '../documents';
import { fixtures } from './fixtures';
import instrumentRoutes from '../../utils/instrumentRoutes';
import templates from '../../templates';

describe('documents', () => {
  let routes;

  beforeEach(done => {
    db.clearAllAndLoad(fixtures)
      .then(done)
      .catch(catchErrors(done));
    routes = instrumentRoutes(documentRoutes);
  });

  afterAll(done => {
    db.disconnect().then(done);
  });

  describe('POST', () => {
    let req;

    beforeEach(() => {
      req = {
        body: { title: 'Batman begins' },
        user: { _id: db.id(), username: 'admin' },
        language: 'es',
      };
    });

    it('should need authorization', () => {
      spyOn(documents, 'save').and.returnValue(new Promise(resolve => resolve('document')));
      expect(routes.post('/api/documents', req)).toNeedAuthorization();
    });

    it('should create a new document with current user', done => {
      spyOn(documents, 'save').and.returnValue(new Promise(resolve => resolve('document')));
      routes
        .post('/api/documents', req)
        .then(document => {
          expect(document).toBe('document');
          expect(documents.save).toHaveBeenCalledWith(req.body, {
            user: req.user,
            language: req.language,
          });
          done();
        })
        .catch(catchErrors(done));
    });
  });

  describe('/api/documents', () => {
    beforeEach(() => {
      spyOn(documents, 'getById').and.returnValue(new Promise(resolve => resolve('documents')));
    });
    it('should have a validation schema', () => {
      expect(routes.get.validation('/api/documents')).toMatchSnapshot();
    });
    it('should return documents.get', done => {
      const req = { query: { _id: 'id' }, language: 'es' };
      routes
        .get('/api/documents', req)
        .then(response => {
          expect(documents.getById).toHaveBeenCalledWith(req.query._id, req.language);
          expect(response).toEqual({ rows: ['documents'] });
          done();
        })
        .catch(catchErrors(done));
    });
  });

  describe('/api/documents/count_by_template', () => {
    beforeEach(() => {
      spyOn(templates, 'countByTemplate').and.returnValue(new Promise(resolve => resolve(2)));
    });
    it('should have a validation schema', () => {
      expect(routes.get.validation('/api/documents/count_by_template')).toMatchSnapshot();
    });
    it('should return count of documents using a specific template', done => {
      const req = { query: { templateId: 'templateId' } };

      routes
        .get('/api/documents/count_by_template', req)
        .then(response => {
          expect(templates.countByTemplate).toHaveBeenCalledWith('templateId');
          expect(response).toEqual(2);
          done();
        })
        .catch(catchErrors(done));
    });
  });

  describe('DELETE', () => {
    beforeEach(() => {
      spyOn(documents, 'delete').and.returnValue(Promise.resolve({ json: 'ok' }));
    });

    it('should have a validation schema', () => {
      expect(routes.delete.validation('/api/documents')).toMatchSnapshot();
    });

    it('should use documents to delete it', done => {
      const req = { query: { sharedId: 123, _rev: 456 } };
      return routes
        .delete('/api/documents', req)
        .then(() => {
          expect(documents.delete).toHaveBeenCalledWith(req.query.sharedId);
          done();
        })
        .catch(catchErrors(done));
    });
  });
});
