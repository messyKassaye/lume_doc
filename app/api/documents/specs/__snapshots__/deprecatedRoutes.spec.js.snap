// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documents /api/documents should have a validation schema 1`] = `
Object {
  "children": Object {
    "_id": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "type": "object",
}
`;

exports[`documents /api/documents/count_by_template should have a validation schema 1`] = `
Object {
  "children": Object {
    "templateId": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "rules": Array [
        Object {
          "arg": Object {
            "pattern": /\\^\\[0-9a-fA-F\\]\\{24\\}\\$/,
          },
          "name": "regex",
        },
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`documents DELETE should have a validation schema 1`] = `
Object {
  "children": Object {
    "sharedId": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;
